### PR TITLE
Support unboxed ints in mixed blocks

### PR DESCRIPTION
**Pull request by ncik-roberts**
This PR is about adding constructors to a variant. Those constructors: `Bits32`, `Bits64`, and `Word`. That variant: the unboxed types allowed in mixed blocks.

This PR involves changes in similar parts as #2508. Those are:
  * the backend, to emit code for accessing unboxed int64/nativeint/int32 fields (beware: we can't reuse the unboxed int array access functions, which have an extra word at the start of the block)
  * the front-end, to allow bits32/bits64/word layouts in blocks, and
  * the middle-end, to faithfully plumb through what the front-end desired

Have you heard of "boiling the frog"? This PR realizes a temperature change (i.e. increase in number of constructors) substantial enough to encourage me to do some small cleanups and consolidations of matches on the relevant variant. The frog escapes.

This PR is also about reorganizing and adding tests.
  * **Step 1: reorganize**. Previously `mixed-blocks/test_mixed_blocks.ml` was a mix of "tests for floats in mixed blocks" and "tests for mixed blocks in general". I separate out the "tests for floats in mixed blocks" into `typing-layouts-float64`, to allow a cleaner template for copying in the other layouts' directories.
  * **Step 2: reorganize some more**. Pull out some per-layout `basics.ml` tests into `basics_alpha.ml`/`basics_beta.ml` to account for the fact that mixed blocks aren't yet allowed in stable
  * **Step 3: add tests**. Copy the template of `typing-layouts-float64/unboxed_floats_alpha.ml` for all of the types we've added to mixed blocks. This ensures that several common sources of bugs are well tested (projecting out things then putting them into closures; recursive groups; math).

**Review**
  * @ccasin: Please review the typing changes and the shape of the tests changes.
  * @TheNumbat + @Gbury : Please review the middle-end changes and the backend changes. The middle-end changes are straightforward. The backend's trickiness is concentrated in `backend/cmm_helpers.ml`, which deserves more scrutiny.

### CODE REVIEW

File: backend/cmm_helpers.ml [hunk 1]
```diff
@@ -1246,6 +1246,41 @@ let get_field_computed imm_or_ptr mutability ~block ~index dbg =
   Cop
     (Cload { memory_chunk; mutability; is_atomic = false }, [field_address], dbg)
 
+(* Getters for unboxed int fields *)
+
+let get_field_unboxed_int32 mutability ~block ~index dbg =
+  (* CR mixed blocks: Could a backend person check the "signed"? *)
```
Comment (TheNumbat) [inline backend/cmm_helpers.ml:1252]: Yeah, unboxed int32 memory operations should use signed. (The difference is that loading a 32-bit memory location into a 64-bit register sign-extends.)

Comment (TheNumbat) [inline backend/cmm_helpers.ml:1252]: Yeah, unboxed int32 memory operations should use signed. (The difference is that loading a 32-bit memory location into a 64-bit register sign-extends.)

**Review by TheNumbat — COMMENTED** [review]
`backend/` and `middle_end/` changes look good.

**Review by ccasin — APPROVED** [review]
Looks good - once again only minor suggestions.

File: ocaml/testsuite/tests/mixed-blocks/constructor_args.ml [hunk 2]
```diff
@@ -35,24 +29,24 @@ type t =
...
-  | Mixed6 of float * int * float#
-  | Mixed7 of float * int * float# * int
-  | Mixed8 of float * int * float# * int * float#
-  | Uniform2 of float * float * int
+  | Mixed4 of float * float# * int32#
+  | Mixed5 of float * float# * int64# * float#
+  | Mixed6 of float * int32# * float#
+  | Mixed7 of float * int64# * float# * nativeint#
+  | Mixed8 of float * int32# * float# * int64# * float#
+  | Uniform2 of float * float
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/testsuite/tests/mixed-blocks/constructor_args.ml:37]: Indeed, that seems like a good thing to keep around. (Especially for value slots of a closure.) I've adjusted it.

File: ocaml/testsuite/tests/typing-layouts-float64/unboxed_floats_alpha.ml [hunk 3]
```diff
@@ -220,3 +220,196 @@ let _ =
...
+type t17 = { a : float;
+            dummy : string;
+             mutable b : int;
+             c : float#;
+             mutable d : float#;
+             e : int;
+             mutable f : float# }
+
+(* Construction *)
+let t17_1 = { a = 3.17;
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-float64/unboxed_floats_alpha.ml:315]: Yes, I used find-and-replace extensively so I'm not surprised I misfired here. (I think in an intermediate step I renamed this test to 14, but ended up needing it to be 17?) I double-checked the moved code to confirm this was the only change.

[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-float64/unboxed_floats_alpha.ml:315]: Thanks for your vigilance.

File: ocaml/testsuite/tests/typing-layouts-word/basics_beta.ml [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+[%%expect{|
+type ('a : word) t5_7 = A of int
+Line 2, characters 24-31:
+2 | type ('a : word) t5_8 = A of 'a;;
+                            ^^^^^^^
+Error: The enabled layouts extension does not allow for mixed constructors.
+       You must enable -extension layouts_alpha to use this feature.
+|}]
+
+type t11_1 = ..
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-word/basics_beta.ml:90]: thanks, fixed in both.

File: ocaml/typing/typedecl.ml [hunk 5]
```diff
@@ -1098,7 +1099,7 @@ let check_abbrev env sdecl (id, decl) =
    should be replaced with checks at the places where values of those types are
    constructed.  We've been conservative here in the first version. This is the
    same issue as with arrows. *)
-let check_representable ~why ~allow_float env loc kloc typ =
+let check_representable ~why ~allow_unboxed env loc kloc typ =
   match Ctype.type_sort ~why env typ with
   (* CR layouts v3: This is a convenient place to rule out [float#] in
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/typing/typedecl.ml]: good catch, thanks.

File: ocaml/testsuite/tests/typing-layouts-bits32/basics_beta.ml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+|}];;
+
+(* Test 5: structures *)
+
+type t5_1 = { x : t_bits32 };;
+[%%expect{|
+Line 1, characters 0-28:
+1 | type t5_1 = { x : t_bits32 };;
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The enabled layouts extension does not allow for mixed records.
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-bits32/basics_beta.ml:24]: Right now, "mixed record" means any record that's compiled to a "mixed block", where a "mixed block" is something whose header indicates that there's a non-scannable suffix. The name "mixed" here might raise eyebrows if the value prefix is length zero but I think it's a fine state.

I don't have immediate plans to rework these names and concepts to single out all-non-value records differently. And I'm not actively trying to think of a better all-encompassing name for mixed blocks. (Note we can't use abstract tag here as we might be, say, in the third constructor of a variant.)

[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/testsuite/tests/typing-layouts-bits32/basics_beta.ml:24]: I'm resolving this for now. I suppose "unscanned record" might be a good name to put into error messages, but I'm reluctant to push this naming all the way through the compiled ("mixed or unscanned block" sounds terrible).

File: backend/cmm_helpers.ml [hunk 7]
```diff
@@ -1246,6 +1246,40 @@ let get_field_computed imm_or_ptr mutability ~block ~index dbg =
...
+(* Setters for unboxed int fields *)
+
+let setfield_unboxed_int32 arr ofs newval dbg =
+  (* CR layouts v5.1: We will need to vary log2_size_addr when int32 fields are
+     efficiently packed. *)
+  return_unit dbg
+    (Cop
+       ( Cstore (Thirtytwo_signed, Assignment),
+         [array_indexing log2_size_addr arr ofs dbg; newval],
+         dbg ))
```
Comment (Gbury) [inline backend/cmm_helpers.ml:1274]: same remark as above the address of the load should probably be different on big-endian/little-endian architectures (or there should be an assertion+comment to state that we only care about one architecture)

File: middle_end/flambda2/to_cmm/to_cmm_primitive.ml [hunk 8]
```diff
@@ -107,7 +107,12 @@ let block_load ~dbg (kind : P.Block_access_kind.t) (mutability : Mutability.t)
   | Mixed { field_kind = Flat_suffix field_kind; _ } -> (
     match field_kind with
     | Imm -> C.get_field_computed Immediate mutability ~block ~index dbg
-    | Float | Float64 -> C.unboxed_float_array_ref block index dbg)
+    | Float | Float64 ->
+      (* CR layouts v5.1: We should use the mutability here. *)
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_primitive.ml:111]: Maybe clarify the comment a bit: the current code behaves as if this read is always mutable, but when it is immutable we could potentially generate different/better code, if I understand correctly.

**Review by Gbury — CHANGES_REQUESTED** [review]
Ok for the cmm_helpers/to_cmm parts (modulo the comment about endianness).

File: backend/cmm_helpers.ml (hunk 7 above)
Comment (ncik-roberts) [reply backend/cmm_helpers.ml:1285]: I've added an assertion. It shouldn't be hard to add support for big endian as-needed in the future.

File: middle_end/flambda2/to_cmm/to_cmm_primitive.ml (hunk 8 above)
Comment (ncik-roberts) [reply middle_end/flambda2/to_cmm/to_cmm_primitive.ml]: Good idea, done.

**Review by Gbury — APPROVED** [review]
Approving the cmm_helpers/to_cmm parts.

**Pull request merged**